linked_list/linked_list.py

```diff
@@ -86,24 +145,45 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+        if self.head == None:
+            return None
+
+        current = self.head
+        previous = None
+
+        while current is not None:
+            next = current.next
+            current.next = previous
+            previous = current
+            current = next
+        self.head = previous    
+        return self.head
+    
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def find_middle_value(self):
-        pass
+        middle_of_list = int(self.length()/2)
+        middle_value = self.get_at_index(middle_of_list)
+        return middle_value
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
     # Time Complexity: ?
     # Space Complexity: ?
     def find_nth_from_end(self, n):

```

Comment:
👍 Good use of arithmetic and `length`, but what about the time/space complexity?